### PR TITLE
Vr modifiers addition

### PR DESCRIPTION
**Pull request by QuentinDaveau**
Added a new ModifiersManager class which manages the VR modifiers. These modifiers are setting the main controller used in the game, enabling both controllers to be used at the same time, a mirror effect mirroring the controllers' positions and rotations, a prism offset shifting the view angle, and an eye patch effect to disable a given eye. A test class is currently implemented to call the different functions through keyboard and mouse input. Some minor bugs from the code overhaul also have been fixed.

### CODE REVIEW

**Comment by QuentinDaveau** [comment]
Referencing #1 and #14

**Comment by YoNeXia** [comment]
It is good for me!

I had some questions about the code, and Quentin explained me.
Commit descriptions are OK.

I think he can squash commits, but Quentin explained me why he did not, and finally I am agree with him.

Your turn now Bastian! 👊

**Comment by bastianilso** [comment]
Thanks @YoNeXia ! 👍 

> I had some questions about the code, and Quentin explained me.
> Commit descriptions are OK.
which questions? Would be great if you could share those here - it's likely that I will end up asking Quentin very similar questions. 😉 

@QuentinDaveau review coming up asap.. 🎈

File: Assets/Scripts/ModifiersManager.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        SetMainController(true);
+    }
+
+    // Sets an eye patch. -1 = left eye, 0 = none and 1 = right eye. Calls WaitForCameraAndUpdate coroutine to set eye patch.
+    public void SetEyePatch(int value)
+    {
+        if (value < -1 || value > 1) return;
+        eyePatch = value;
+        StartCoroutine(WaitForCameraAndUpdate(eyePatch));
+    }
```
[the remark this answers is not in the document]
Comment (QuentinDaveau) [reply Assets/Scripts/ModifiersManager.cs]: Good point ! I'll fix that.

[the remark this answers is not in the document]
Comment (QuentinDaveau) [reply Assets/Scripts/ModifiersManager.cs]: Fixed.

**Comment by QuentinDaveau** [comment]
Warning for the eye patch added at the beginning of ModifiersManager and at the function call. I think it should be all good for merging.

File: Assets/Scripts/ControllerModifierManager.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    // On VR update, update the position and rotation so they are mirrored
+    private void OnPoseUpdated(bool obj)
+    {
+        Vector3 controllerLocalPosition = controllerPose.GetLocalPosition(inputSource);
+        Vector3 controllerToHead = controllerLocalPosition - cameraTransform.localPosition;
+        Vector3 controllerAngles = controllerPose.GetLocalRotation(inputSource).eulerAngles;
+
+        // Update the position and rotation using the wall y axis as mirroring plane
+        transform.localPosition = cameraTransform.localPosition + Vector3.Reflect(controllerToHead, Quaternion.AngleAxis(wallTransform.localEulerAngles.y, Vector3.up) * Vector3.right);
+        transform.localRotation = Quaternion.Euler(new Vector3(controllerAngles.x, (2 * wallTransform.eulerAngles.y) - controllerAngles.y , -controllerAngles.z));
```
Comment (bastianilso) [inline Assets/Scripts/ControllerModifierManager.cs]: For sake of readability, could you split your arithmethic operations up into several lines and store the intermediate results in variables with sensible names (and comments)? When we introduce more complicated arithmethics, we need to document them, so they are easy to debug later.

File: Assets/Scripts/ModifiersManager.cs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+    // Sets the eye patch. Forces the camera to render a black screen for a short duration and disables an eye while the screen is black.
+    // If the image wasn't forced black we would have a frozen image of the game in the disabled eye.
+
+    /*
+    WARNING: Due to the Vive overlay, it is necessary to disable the chaperone bounds in the Vive settings, otherwise the eye patch would still render the chaperone if near a boundary.
+    To do so, go the the VR settings WHILE INSIDE THE HEADSET -> Chaperone -> select DEVELOPER MODE and set a neutral color with the lowest opacity possible.
+    It is also possible to fully hide the chaperone by editing the steamvr.vrsettings file and setting "CollisionBoundsColorGammaA" to 0.
+    */
```
Comment (bastianilso) [inline Assets/Scripts/ModifiersManager.cs]: This might also be good to have in README.md, in a FAQ section.

**Comment by QuentinDaveau** [comment]
The changes in Pointer.cs are to be able to Enable/Disable a pointer on the fly. It is related to the ModifiersManager since it uses this feature to set the main controller and the dual task mode at runtime. I can put this change in another commit if you want.

**Comment by bastianilso** [comment]
> 
> 
> The changes in Pointer.cs are to be able to Enable/Disable a pointer on the fly. It is related to the ModifiersManager since it uses this feature to set the main controller and the dual task mode at runtime. I can put this change in another commit if you want.

Thanks for the explanation! No, that's fine. Just update your commit body text with that explanation.

**Pull request merged**